### ISSUE TITLE
loadScript has option to also create JSAPI css link tag(s)

### ISSUE THREAD

**Issue by jwasilgeo** [Esri/esri-loader#6]
How about getting the `bootstrap` method to _also_ generate a CSS `<link>` tag when it creates the JS `<script>` tag?  Could lead to fewer steps for consuming developers to have to remember to deal with.

https://github.com/tomwayson/esri-loader/blob/master/index.ts#L20

Perhaps the `options` could be extended to include not only the JS url but also the CSS url? This would be helpful since the CSS url has different structures throughout the different incarnations of JSAPI 3.x and 4.x. Something like:
`options.cssUrl = 'https://js.arcgis.com/4.2/esri/css/main.css';`

**Comment by jwasilgeo** [comment]
Actually, on 2nd thought, how about making the `options` property for css be an array?  If someone is using JSAPI 3.x but wants the optional calcite theme, they need to pull in 2 css links.

https://developers.arcgis.com/javascript/3/jshelp/css.html

**Comment by tomwayson** [comment]
Good idea @jwasilgeo!

A few thoughts:
- Suggest adding an `includeCss` option and default it to `false` so this would not be a breaking change
- I would prefer a `version` option (`version: 3.19`) that was able to construct the JS and CSS urls (I've wanted to do that for the JS anyway)
- perhaps a `theme: 'calcite'` option to handle the themes
- could make both of the above work for non-CDN builds w/ something like a `baseUrl` option, which defaults to CDN base URL

If we did all that I think we could deprecate the existing `url` option and would not need a `cssUrls` option, right?

What do you think @davetimmins?

**Comment by tomwayson** [comment]
Would `isLoaded()` also need to check if the CSS is loaded? If so, should it return different states (i.e. `'script'`, `'styles'`, `'all'` or `false`)? Would it be only if someone passed `includeCss`, or even if someone loaded the CSS manually too?

**Comment by davetimmins** [comment]
Personally I prefer being more explicit about what is being loaded. I like the idea of having a version option which loads all the required assets but for setting non CDN or other assets I'd use separate arrays for the asset types e.g.

```js
{
  version: 3.19
}
// or
{
  js: ['https://myserver.com/esri-js/'],
  css: ['https://myserver.com/esri-js/css/main.css', 'https://myserver.com/esri-js/css/themes/calcite.css']
}
```

**Comment by tomwayson** [comment]
Good point @davetimmins. .

So scrap the idea of a `baseUrl` then and just use the `version` for CDN. I'm amenable to that.

**Comment by jwasilgeo** [comment]
If we go with the `version` and try to auto-generate all the magic, my only concern is that for versions < 3.11 the url structure with the CDN is a little different (see first section in https://developers.arcgis.com/javascript/3/jshelp/css.html).

Perhaps I should ask for clarification.  Is the `version` just going to be intended for the JS script tag?  And then people need to specify the css urls separately (e.g. calcite theme, extra dojo stuff if they are using dijits with claro, etc.)?

**Comment by tomwayson** [comment]
No, this version would generate css urls as well if `includeCss: true`. Yes, there are different folder structures for < 3.11. All the more reason to abstract this logic away from the user. Nothing a few unit tests couldn't handle.

**Comment by jwasilgeo** [comment]
Alright, can y'all edit this comment (or add your own comments) to document the full set of desired `options` properties so we're all on the same page?  Thanks.


```javascript
{
  version: '4.2', // or '3.19compact', '3.10', etc.
  includeCss: false // generate css urls if true
}
```

The desired output will be 1 script tag and 1+ link css tag(s) added to the DOM. (right?)

**Comment by tomwayson** [comment]
Or we could do some elegant thing w/ a `css` option that would replace `includeCss` and `cssUrls`:
 - falsey means don't do any css loading
 - `true` means figure out default path based on `version` or `url` and include
 - `true` plus `theme: 'calcite'` means figure out default path to theme and include that too
 - an `Array` means load each as a url

I kind of like that better.

**Comment by jwasilgeo** [comment]
How are we feeling about this topic now that there’s been quite a few changes to the API? Is this something we still want to entertain for something like the `loadScript` method? I could be swayed either way.

**Comment by tomwayson** [comment]
There hasn't been overwhelming demand.

I would certainly accept a PR that implements this ad described in https://github.com/Esri/esri-loader/issues/6#issuecomment-279531108 especially if it has tests.

**Comment by tomwayson** [comment]
I should have said "accept a PR that has the minimum viable implementation that can be iterated on to achieve the API described above."

And looking at it now, I think that list is missing the simplest thing: 
- `string` treat that as a url and load it

From there we could then support `Array` in a follow on PR.

Then maybe `true` and `theme: 'calcite'` in later PRs. Someday. Maybe. I think URL parsing is going to be hard - easy enough for the CDN versions, but hard for locally hosted builds. So even `true` could be handled incrementally. It might only support CDN URLs at first.

**Comment by jwasilgeo** [comment]
Continuing with both this discussion and with PR #80, I suppose we first need to revisit these questions:

- Should this repo even provide the (optional) functionality to load in the Esri CSS dependency?

- If yes, should it be automatic (i.e. optional, but perhaps default ON)?

My personal opinion is to enable this functionality so that esri-loader would take care of all the Esri dependencies for consuming developers to get up and running as quickly as possible.  (where "all" means 1 JS resource and at least 1 CSS resource)

With that said though, I can also be swayed to see the other side of the argument.  Adding a CSS `<link>` to HTML by hand is straightforward, whereas _integrating the Esri JS_ and _documenting best practices_ is where esri-loader really shines and also keeps its own footprint as small as possible.

To muddy the waters as much as possible, here are some comparisons:

mapping lib | dependencies
--- | ---
Esri JSAPI | JS and CSS are required
Leaflet | JS and CSS are required
OpenLayers | JS and CSS are required
Google Maps | JS only is required (`<style>` is appended to DOM automatically) 

And we already know that there are different variations in both content and url structure of the Esri CSS stylesheet(s) depending on which specific version of v3 or v4 you're using.

Now I don't know anymore how to proceed, but if we do this then I'd push for the implementation that adds the least amount of new source code and doesn't introduce non-optional API features.

**Comment by jwasilgeo** [comment]
JSAPI version | css notes
--- | ---
3.2 to 3.10 | `https://js.arcgis.com/<VERSION>/js/esri/css/esri.css`*
3.11 to latest | `https://js.arcgis.com/<VERSION>/esri/css/esri.css`* 
3.16 to latest (theme **replacement** css) |  `https://js.arcgis.com/<VERSION>/esri/themes/calcite/dijit/calcite.css` AND `https://js.arcgis.com/<VERSION>/esri/themes/calcite/esri/esri.css`
4.0 to latest | `https://js.arcgis.com/<VERSION>/esri/css/main.css` OR `https://js.arcgis.com/<VERSION>/esri/css/view.css`
4.0 to latest (theme **replacement** css) |  `https://js.arcgis.com/<VERSION>/esri/themes/<THEME_NAME>/main.css` where THEME_NAME can be: `light`, `dark`, `light-blue`, `dark-blue`, `light-green`, `dark-green`, `light-purple`, `dark-purple`, `light-red`, OR `dark-red`

*_and optional dojo themes and widgets css_

gathered from:
- https://developers.arcgis.com/javascript/3/jshelp/css.html
- https://developers.arcgis.com/javascript/latest/guide/styling/index.html

**Comment by andygup** [comment]
@tomwayson in the default `loadModules` use case where it uses the latest JS API version, would it also make sense to have `loadModules` automatically inject the latest css as well?

For example, the library could specify it as `DEFAULT_CSS_URL` similar to  https://github.com/Esri/esri-loader/blob/master/src/esri-loader.ts#L14. Or it could simply infer the css url pattern.

That would eliminate the need to do something like this and you could simply just use `loadModules()` without `options`:

```
 const options = {
      css: 'https://js.arcgis.com/4.6/esri/css/main.css'
    };
    loadModules([
      'esri/Map',
      'esri/views/MapView'
    ], options).then(([EsriMap, EsriMapView]) => {

```

Thoughts?

/cc @jwasilgeo

**Comment by tomwayson** [comment]
Thanks @andygup

That's pretty much _exactly_ what https://github.com/Esri/esri-loader/pull/80 would have done. You can read the comments there on why I rejected it, but TLDR: I don't think it's a good idea for this library to load the CSS by default.

The API that I thinks makes the most sense is for options to allow `{ css: true }` (#88). I like that b/c:
- it's opt in
- but doesn't need to be kept in sync w/ the default script URL
- supports non-default script URLs (3.x, etc)

@jwasilgeo's above table describes how `{ css: true }` will be have. We can build up support for all those cases over time incrementally w/o any breaking changes. It's currently just a matter of someones time to start writing `getCssUrl(scriptUrl)`, which I think would be easier if I could get to #87, but is not really blocked by that at all. In the mean time, consumes can still use `{ css: 'https://js.arcgis.com/4.6/esri/css/main.css' }`.

**Comment by andygup** [comment]
@tomwayson Thanks for the addt'l background info.

>> injecting styles by default would be a breaking change and require a major version bump.

Makes sense. It would also be bad form to have a duplicate CSS hanging around in the DOM.

On a related note, we may want to document that lazy loading the CSS with this library forces an additional reflow.

**Comment by tomwayson** [comment]
> we may want to document that lazy loading the CSS with this library forces an additional reflow.

Good point @andygup!

I had been thinking that's OK though, b/c I have been interpreting the messaging from Lighthouse/Addy/Ilya as land (even inline) critical CSS as soon as possible, and defer loading other CSS for later as an implicit endorsement of the idea of breaking up your CSS by route. However after a closer read of [Redner blocking CSS](https://developers.google.com/web/fundamentals/performance/critical-rendering-path/render-blocking-css) I see they're mainly talking about using media queries (i.e. print styles, etc). 

That doesn't necessarily mean that you should **not** break up and load your CSS by route. After all, the whole CSS in JS phenomenon effectively has you breaking down you styles by _component_. I don't have a real firm grasp on how that stuff works, but I think a fundamental level it's [either injecting individual style into the DOM as each component loads or dynamically changing the rules of an existing stylesheet](https://medium.com/@oleg008/aphrodite-vs-jss-a15761b91ee3). That stuff has _got_ to be causing constant reflows, right? Maybe it doesn't matter b/c it's at the same time as the DOM is being updated (new components added/removed)?

Another thing is what  about`rel="prefetch"`? Is that better than esri-loader's `loadCss()`? Or is it better to use both - prefetch on non-map routes, `loadCss()` on the map route? Maybe the best thing is for the page to always have a `<link />` but dynamically set `rel="prefetch"` based on whether or not it's the map route? Maybe none of that even matters.

In short, I'm still learning about and experimenting w/ this stuff, and don't feel comfortable advising people what to do, but if you've got a solid grasp on any of this, I'll welcome any PRs w/ such advice.

**Comment by andygup** [comment]
I guess what I'm saying is I do like `loadCSS()` as a helper method for prototyping-only. My concern about `loadCSS()` is it operates outside of Angular standard practices and potentially other frameworks as well. Angular has a ton of built-in methods for dealing with this where styles are applied dynamically or statically as an integral part of its rendering execution pipeline and implemented within the component life-cycle.  

For Angular we can potentially solve the issue by documenting something like: "`loadCSS()` is a simple helper method, however we recommend researching your chosen framework's built-in style loader mechanisms for the best performance." 



For example, in Angular you can add global styles directly into the build by specifying them in`.angular-cli.json > apps > styles` and there's Shadow DOM and Light DOM selectors, encapsulation modes, and even styling specificity.

**Comment by tomwayson** [comment]
As of #163 most of the features discussed in this issue have been implemented. The remaining ones will be handled in new issues (like #157).
